modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,425 @@
+	buffer3    []string          // Last hashed values
+	overflow   uint64
+}
+
+// GetBuffersForDimensions takes advantage of the fact that the configuration for a tracker
+// changes slowly.  Reuses buffers from the previous call when the dimensions are the same.
+func (t *tenantUsage) GetBuffersForDimensions(dimensions map[string]string) ([]mapping, []string, []string, []string) {
+	if !maps.Equal(dimensions, t.dimensions) {
+		// The configuration changed.
+
+		// Step 1
+		// Gather all configured dimensions and their sanitized output
+		t.dimensions = dimensions
+		sanitizedDimensions := make(map[string]string, len(dimensions))
+		for k, v := range dimensions {
+			// Get the final sanitized output label for this
+			// dimension.  Dimensions are key-value pairs with
+			// optional value.  If value is empty string, then
+			// we use the just the key.  Regardless the output
+			// is always the sanitized version.
+			// Example:
+			//    service.name="" 			=> "service_name"
+			//    service.name="foo.bar"	=> "foo_bar"
+			var sanitized string
+			if v == "" {
+				// The dimension is using default mapping
+				v = k
+			}
+			sanitized = strutil.SanitizeFullLabelName(v)
+			sanitizedDimensions[k] = sanitized
+		}
+
+		// Step 2
+		// Build the final list of sorted/distinct outputs
+		t.sortedKeys = t.sortedKeys[:0]
+		for _, v := range sanitizedDimensions {
+			if !slices.Contains(t.sortedKeys, v) {
+				t.sortedKeys = append(t.sortedKeys, v)
+			}
+		}
+		slices.Sort(t.sortedKeys)
+
+		// Step 3
+		// Prepare the mapping from raw attribute names to the final location of
+		// where it goes in the output buffers. This avoids another layer of indirection.
+		t.mapping = t.mapping[:0]
+		for k := range dimensions {
+			i := slices.Index(t.sortedKeys, sanitizedDimensions[k])
+			t.mapping = append(t.mapping, mapping{
+				from: k,
+				to:   i,
+			})
+		}
+
+		// Step 4
+		// Prepopulate the buffers and precompute the overflow bucket
+		t.buffer1 = make([]string, len(t.sortedKeys))
+		t.buffer2 = make([]string, len(t.sortedKeys))
+		t.buffer3 = make([]string, len(t.sortedKeys))
+		for i := range t.sortedKeys {
+			t.buffer1[i] = overflowLabel
+		}
+		t.overflow = hash(t.sortedKeys, t.buffer1)
+	}
+	return t.mapping, t.buffer1, t.buffer2, t.buffer3
+}
+
+// func (t *tenantUsage) getSeries(labels, values []string, maxCardinality uint64) *bucket {
+func (t *tenantUsage) getSeries(buffer []string, maxCardinality uint64) *bucket {
+	h := hash(t.sortedKeys, buffer)
+
+	b := t.series[h]
+	if b == nil {
+		// Before creating a new series, check for cardinality limit.
+		if uint64(len(t.series)) >= maxCardinality {
+			// Overflow
+			// This tenant is at the maximum number of series.  In this case all data
+			// goes into the final overflow bucket. It has the same dimensions as the
+			// current configuration, except every label is overridden to the special overflow value.
+			for k := range buffer {
+				buffer[k] = overflowLabel
```

Comment:
calling code is ok with the passed buffer being overwritten?

Reply:
It is, and actually requires it to ensure that the caller optimization to skip unnecessary bucket rehashing works.   Looking at it now, this area could be simplified a bit more.  For example by pre-creating the overflow bucket, and hiding the complexity of the rehashing by moving it around.  However I'm good with the code in its current state.